primitives/state-machine/src/overlayed_changes.rs

```diff
@@ -223,11 +223,28 @@ impl OverlayedChanges {
 			})
 	}
 
-	/// Returns mutable reference to either commited or propsective value.
-	pub fn value_mut(&mut self, key: &[u8]) -> Option<&mut Option<StorageValue>> {
-		// not using map because of double borrow inside closure
-		if let Some(entry) = self.prospective.top.get_mut(key) { return Some(&mut entry.value) }
-		return self.committed.top.get_mut(key).map(|entry| &mut entry.value);
+	/// Returns mutable reference to current changed value (prospective).
+	/// If there is no value in the overlay, the default callback is used to initiate
+	/// the value.
+	/// Warning this function register a change, so the mutable reference MUST be modified.
+	pub fn value_mut_or_insert_with(
+		&mut self,
+		key: &[u8],
+		init: impl Fn() -> StorageValue,
+	) -> &mut StorageValue {
+		let extrinsic_index = self.extrinsic_index();
+		let committed = &self.committed.top;
+		let entry = self.prospective.top.entry(key.to_vec())
```

Comment:
```suggestion
		match self.prospective.top.entry(key.to_vec()) {
			Entry::Vacant(vacant) => {
			  vacant.insert();
				if let Some(extrinsic) = extrinsic_index {
			entry.extrinsics.get_or_insert_with(Default::default)
				.insert(extrinsic);
		}
		if entry.value.is_none() {
			entry.value = Some(init());
		}
			},
			Entry::Occupied(occupied) => occupied.get_mut(),
		}
```

Something in this direction.

Reply:
Damn the code is not correct (if there is none in committed we would have overwrite it with backend value).

Reply:
Ok, I fix that, @bkchr , I am not sure what we try to achieve by matching (when I write it with matching it ends back to the older version because ther is nothing specific to occupied)

Reply:
Maybe not updating change extrinsic info? (could also do it after append, by exposing another function)